File: vsg_qt/track_settings_dialog/ui.py
from __future__ import annotations

from PySide6.QtWidgets import (
    QCheckBox,
    QComboBox,
    QDialog,
    QDialogButtonBox,
    QDoubleSpinBox,
    QFormLayout,
    QGroupBox,
    QLineEdit,
    QPushButton,
    QVBoxLayout,
)

from .logic import TrackSettingsLogic


class TrackSettingsDialog(QDialog):
    """Small popup dialog to edit per-track options."""
    def __init__(self, track_type: str, codec_id: str, track_data: dict = None, **kwargs):
        super().__init__()
        self.setWindowTitle("Track Settings")
        self.setMinimumWidth(400)
        self.track_data = track_data or {}

        # --- UI Elements ---
        # Language selector (for all track types)
        self.lang_combo = QComboBox()

        # Custom track name (for all track types)
        self.custom_name_input = QLineEdit()

        # Subtitle-specific controls
        self.cb_ocr = QCheckBox("Perform OCR")
        self.cb_convert = QCheckBox("Convert to ASS (SRT only)")
        self.cb_rescale = QCheckBox("Rescale to video resolution")
        self.size_multiplier = QDoubleSpinBox()
        self.size_multiplier.setRange(0.1, 10.0)
        self.size_multiplier.setSingleStep(0.1)
        self.size_multiplier.setDecimals(2)
        self.size_multiplier.setPrefix("Size multiplier: ")
        self.size_multiplier.setSuffix("x")

        # Frame sync exclusions button (ASS/SSA only)
        self.sync_exclusion_btn = QPushButton("Configure Frame Sync Exclusions...")
        self.sync_exclusion_btn.clicked.connect(self._open_sync_exclusion_dialog)

        # --- Logic ---
        self._logic = TrackSettingsLogic(self)

        # --- Layout ---
        layout = QVBoxLayout(self)

        # Language section (always visible)
        lang_group = QGroupBox("Language Settings")
        lang_layout = QFormLayout(lang_group)
        lang_layout.addRow("Language Code:", self.lang_combo)
        layout.addWidget(lang_group)

        # Track name section (always visible)
        name_group = QGroupBox("Track Name")
        name_layout = QFormLayout(name_group)
        name_layout.addRow("Custom Name:", self.custom_name_input)
        layout.addWidget(name_group)

        # Subtitle section (conditionally visible)
        self.subtitle_group = QGroupBox("Subtitle Options")
        subtitle_layout = QVBoxLayout(self.subtitle_group)
        subtitle_layout.addWidget(self.cb_ocr)
        subtitle_layout.addWidget(self.cb_convert)
        subtitle_layout.addWidget(self.cb_rescale)
        subtitle_layout.addWidget(self.size_multiplier)
        subtitle_layout.addWidget(self.sync_exclusion_btn)
        layout.addWidget(self.subtitle_group)

        btns = QDialogButtonBox(QDialogButtonBox.Ok | QDialogButtonBox.Cancel)
        btns.accepted.connect(self.accept)
        btns.rejected.connect(self.reject)
        layout.addWidget(btns)

        # --- Initial State ---
        self._logic.apply_initial_values(**kwargs)
        self._logic.init_for_type_and_codec(track_type, codec_id)

    def _open_sync_exclusion_dialog(self):
        """Open the sync exclusion configuration dialog."""
        from vsg_qt.sync_exclusion_dialog import SyncExclusionDialog

        # Get existing config if available
        existing_config = None
        if self.track_data.get('sync_exclusion_styles'):
            existing_config = {
                'mode': self.track_data.get('sync_exclusion_mode', 'exclude'),
                'styles': self.track_data.get('sync_exclusion_styles', []),
            }

        dialog = SyncExclusionDialog(
            track_data=self.track_data,
            existing_config=existing_config,
            parent=self
        )

        if dialog.exec():
            config = dialog.get_exclusion_config()
            if config:
                # Store in track_data temporarily (will be saved when main dialog is accepted)
                self.track_data['sync_exclusion_styles'] = config['styles']
                self.track_data['sync_exclusion_mode'] = config['mode']
                self.track_data['sync_exclusion_original_style_list'] = config['original_style_list']
            else:
                # Clear exclusions if None returned
                self.track_data.pop('sync_exclusion_styles', None)
                self.track_data.pop('sync_exclusion_mode', None)
                self.track_data.pop('sync_exclusion_original_style_list', None)

    def read_values(self) -> dict:
        """Public method to retrieve the dialog's current values."""
        values = self._logic.read_values()

        # Include sync exclusion config in the returned values
        values['sync_exclusion_styles'] = self.track_data.get('sync_exclusion_styles', [])
        values['sync_exclusion_mode'] = self.track_data.get('sync_exclusion_mode', 'exclude')
        values['sync_exclusion_original_style_list'] = self.track_data.get('sync_exclusion_original_style_list', [])

        return values

File: qt_ui/track_settings_dialog/ui.cpp
// Track Settings Dialog Implementation

#include "ui.hpp"
#include "logic.hpp"

#include <QVBoxLayout>
#include <QFormLayout>
#include <QPushButton>
#include <QDialogButtonBox>
#include <QMessageBox>

TrackSettingsDialog::TrackSettingsDialog(
    const QString& trackType,
    const QString& codecId,
    const std::map<QString, QVariant>& initialValues,
    QWidget* parent)
    : QDialog(parent)
    , m_trackType(trackType)
    , m_codecId(codecId)
{
    setWindowTitle("Track Settings");
    setMinimumWidth(400);

    buildUi();

    m_logic = std::make_unique<TrackSettingsLogic>(this);
    m_logic->applyInitialValues(initialValues);

    initForTypeAndCodec(trackType, codecId);
}

TrackSettingsDialog::~TrackSettingsDialog() = default;

void TrackSettingsDialog::buildUi()
{
    auto* layout = new QVBoxLayout(this);

    // === Language section ===
    auto* langGroup = new QGroupBox("Language Settings");
    auto* langLayout = new QFormLayout(langGroup);

    m_langCombo = new QComboBox();
    m_langCombo->setEditable(true);
    // Add common languages
    m_langCombo->addItem("English", "eng");
    m_langCombo->addItem("Japanese", "jpn");
    m_langCombo->addItem("Spanish", "spa");
    m_langCombo->addItem("French", "fra");
    m_langCombo->addItem("German", "deu");
    m_langCombo->addItem("Chinese", "chi");
    m_langCombo->addItem("Korean", "kor");
    m_langCombo->addItem("Undetermined", "und");

    langLayout->addRow("Language Code:", m_langCombo);
    layout->addWidget(langGroup);

    // === Track name section ===
    auto* nameGroup = new QGroupBox("Track Name");
    auto* nameLayout = new QFormLayout(nameGroup);

    m_customNameInput = new QLineEdit();
    m_customNameInput->setPlaceholderText("Leave blank to use default");

    nameLayout->addRow("Custom Name:", m_customNameInput);
    layout->addWidget(nameGroup);

    // === Subtitle options section ===
    m_subtitleGroup = new QGroupBox("Subtitle Options");
    auto* subtitleLayout = new QVBoxLayout(m_subtitleGroup);

    m_cbOcr = new QCheckBox("Perform OCR (for image-based subtitles)");
    m_cbOcr->setToolTip("Convert PGS/VobSub image subtitles to text using OCR");
    subtitleLayout->addWidget(m_cbOcr);

    m_cbConvert = new QCheckBox("Convert to ASS (for SRT files)");
    m_cbConvert->setToolTip("Convert SRT subtitles to ASS format for styling");
    subtitleLayout->addWidget(m_cbConvert);

    m_cbRescale = new QCheckBox("Rescale to video resolution");
    m_cbRescale->setToolTip("Adjust subtitle positioning for different video resolutions");
    subtitleLayout->addWidget(m_cbRescale);

    m_sizeMultiplier = new QDoubleSpinBox();
    m_sizeMultiplier->setRange(0.1, 10.0);
    m_sizeMultiplier->setSingleStep(0.1);
    m_sizeMultiplier->setDecimals(2);
    m_sizeMultiplier->setValue(1.0);
    m_sizeMultiplier->setPrefix("Size multiplier: ");
    m_sizeMultiplier->setSuffix("x");
    m_sizeMultiplier->setToolTip("Scale subtitle size (1.0 = original size)");
    subtitleLayout->addWidget(m_sizeMultiplier);

    m_syncExclusionBtn = new QPushButton("Configure Frame Sync Exclusions...");
    m_syncExclusionBtn->setToolTip("Exclude certain styles from frame-level sync adjustments");
    connect(m_syncExclusionBtn, &QPushButton::clicked,
            this, &TrackSettingsDialog::onSyncExclusionClicked);
    subtitleLayout->addWidget(m_syncExclusionBtn);

    layout->addWidget(m_subtitleGroup);

    // === Dialog buttons ===
    auto* btns = new QDialogButtonBox(
        QDialogButtonBox::Ok | QDialogButtonBox::Cancel);
    connect(btns, &QDialogButtonBox::accepted, this, &QDialog::accept);
    connect(btns, &QDialogButtonBox::rejected, this, &QDialog::reject);
    layout->addWidget(btns);
}

void TrackSettingsDialog::initForTypeAndCodec(const QString& trackType, const QString& codecId)
{
    // Show/hide subtitle options based on track type
    bool isSubtitle = (trackType == "subtitles");
    m_subtitleGroup->setVisible(isSubtitle);

    if (isSubtitle) {
        QString codec = codecId.toUpper();

        // OCR only available for image-based subtitles
        bool isImageBased = codec.contains("VOBSUB") ||
                            codec.contains("PGS") ||
                            codec.contains("HDMV");
        m_cbOcr->setVisible(isImageBased);

        // Convert to ASS only for SRT
        bool isSrt = codec.contains("UTF8") || codec.contains("SRT");
        m_cbConvert->setVisible(isSrt);

        // Sync exclusion only for ASS/SSA
        bool isAss = codec.contains("ASS") || codec.contains("SSA");
        m_syncExclusionBtn->setVisible(isAss);
    }
}

void TrackSettingsDialog::onSyncExclusionClicked()
{
    // TODO: Open sync exclusion dialog
    QMessageBox::information(this, "Sync Exclusions",
        "Sync exclusion configuration not yet implemented.");
}

std::map<QString, QVariant> TrackSettingsDialog::readValues() const
{
    return m_logic->readValues();
}
